libcontainer/nsenter/nsexec.c

```diff
@@ -192,22 +194,74 @@ static void update_setgroups(int pid, enum policy_t setgroup)
 	}
 }
 
-static void update_uidmap(int pid, char *map, size_t map_len)
+static int attempt_mapping_tool(const char *app, int pid, char *map, size_t map_len)
+{
+#define MAX_ARGV 20
+	char *argv[MAX_ARGV];
+	char pid_fmt[16];
+	int argc = 0;
+	size_t i;
+	int child;
+
+	sprintf (pid_fmt, "%d", pid);
+
+	argv[argc++] = (char *) app;
+	argv[argc++] = pid_fmt;
+	/*
+	 * Convert the map string into a list of argument that newuidmap/newgidmap
+	 * can understand.
+	 */
+	for (i = 0; i < map_len - 1; ) {
+		argv[argc++] = &map[i++];
+
+		while (i < map_len - 1 && map[i] != '\n' && map[i] != ' ' && map[i] != '\0')
+			i++;
+		map[i++] = '\0';
+
+		while (i < map_len - 1 && map[i] == '\n' && map[i] == ' ')
+			i++;
+		if (map[i] == '\0')
+			break;
+	}
+
+	child = fork();
+	if (child < 0)
+		bail("failed to fork");
+
+	if (child == 0) {
+		execv (app, argv);
+	}
+	else {
+		int status;
+		while (waitpid (child, &status, 0) < 0 && errno == EINTR);
```

Comment:
```
do
    waitpid(child, &status, 0);
while (errno == EINTR);
```

If not, just put the `;` on a bare newline so it's more clear it's an empty loop.